lib/private/UserPreferences.php

```diff
@@ -0,0 +1,1777 @@
+	 *
+	 * @since 31.0.0
+	 */
+	public function clearCache(string $userId, bool $reload = false): void {
+		$this->assertParams($userId, allowEmptyApp: true);
+		$this->lazyLoaded[$userId] = $this->fastLoaded[$userId] = false;
+		$this->lazyCache[$userId] = $this->fastCache[$userId] = $this->valueDetails[$userId] = [];
+
+		if (!$reload) {
+			return;
+		}
+
+		$this->loadPreferencesAll($userId);
+	}
+
+	/**
+	 * @inheritDoc
+	 *
+	 * @since 31.0.0
+	 */
+	public function clearCacheAll(): void {
+		$this->lazyLoaded = $this->fastLoaded = [];
+		$this->lazyCache = $this->fastCache = $this->valueDetails = [];
+	}
+
+	/**
+	 * For debug purpose.
+	 * Returns the cached data.
+	 *
+	 * @return array
+	 * @since 31.0.0
+	 * @internal
+	 */
+	public function statusCache(): array {
+		return [
+			'fastLoaded' => $this->fastLoaded,
+			'fastCache' => $this->fastCache,
+			'lazyLoaded' => $this->lazyLoaded,
+			'lazyCache' => $this->lazyCache,
+			'valueDetails' => $this->valueDetails,
+		];
+	}
+
+	/**
+	 * @param int $needle bitflag to search
+	 * @param int $flags all flags
+	 *
+	 * @return bool TRUE if bitflag $needle is set in $flags
+	 */
+	private function isFlagged(int $needle, int $flags): bool {
+		return (($needle & $flags) !== 0);
+	}
+
+	/**
+	 * Confirm the string set for app and key fit the database description
+	 *
+	 * @param string $userId
+	 * @param string $app assert $app fit in database
+	 * @param string $prefKey assert preference key fit in database
+	 * @param bool $allowEmptyUser
+	 * @param bool $allowEmptyApp $app can be empty string
+	 * @param ValueType|null $valueType assert value type is only one type
+	 */
+	private function assertParams(
+		string $userId = '',
+		string $app = '',
+		string $prefKey = '',
+		bool $allowEmptyUser = false,
+		bool $allowEmptyApp = false,
+		?ValueType $valueType = null,
+	): void {
+		if (!$allowEmptyUser && $userId === '') {
+			throw new InvalidArgumentException('userId cannot be an empty string');
+		}
+		if (!$allowEmptyApp && $app === '') {
+			throw new InvalidArgumentException('app cannot be an empty string');
+		}
+		if (strlen($userId) > self::USER_MAX_LENGTH) {
+			throw new InvalidArgumentException('Value (' . $userId . ') for userId is too long (' . self::USER_MAX_LENGTH . ')');
+		}
+		if (strlen($app) > self::APP_MAX_LENGTH) {
+			throw new InvalidArgumentException('Value (' . $app . ') for app is too long (' . self::APP_MAX_LENGTH . ')');
+		}
+		if (strlen($prefKey) > self::KEY_MAX_LENGTH) {
+			throw new InvalidArgumentException('Value (' . $prefKey . ') for key is too long (' . self::KEY_MAX_LENGTH . ')');
+		}
+		if ($valueType !== null) {
+//			$valueFlag = $valueType->value;
+//			if (ValueType::tryFrom($valueFlag) === null) {
+//				throw new InvalidArgumentException('Unknown value type');
+//			}
+		}
+	}
+
+	private function loadPreferencesAll(string $userId): void {
+		$this->loadPreferences($userId, null);
```

Comment:
I think this is a bit confusing, better be explicit:
```suggestion
		$this->loadPreferences($userId, false);
```